README.md
[![GitHub Workflow Status](https://img.shields.io/github/workflow/status/gmaclennan/parse-dms/Node.js%20CI)](https://github.com/gmaclennan/parse-dms/actions?query=workflow%3A%22Node.js+CI%22)
[![npm](https://img.shields.io/npm/v/parse-dms)](https://www.npmjs.com/package/parse-dms)

parseDMS
========

A robust parser for degrees, minutes, seconds latitude and longitude values, with comprehensive tests.

## Demo

[Try it out](http://gmaclennan.github.io/parse-dms) in the browser.

## Usage

```javascript
var parseDMS = require('parse-dms');

parseDMS('59°12\'7.7"N 02°15\'39.6"W')
```

Returns an object with properties `lat` `lon` if it can infer them, or a single number if only one coordinate is provided and it cannot infer lat or lon.

If a DMS coordinate pair is given with no hemisphere letters, then the order is assumed to be lat, lon.

Will throw an error if it cannot parse the string or if degrees, minutes, or seconds are out of range.

Should be able to handle most weird ways that people write DMS values, but if you find one that it cannot handle please submit an issue, or better, submit a pull request with a new test. The regex used is here: https://regex101.com/r/kS2zR1/3

See `test/index.js` for more details on what it can parse.

## Run tests
-----

`npm test`

## Contributing

Pull requests welcome with tests.

## Release History

* 0.0.1 Initial release


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 0.0.5 (2021-02-04)

### 0.0.4 (2021-02-04)


test/index.js
var test = require('tape');
var parseDms = require('../');

test('Correctly parses DMS pairs with different separators, hemisphere at end', function(t) {

    var testData = [
        '59°12\'7.7"N 02°15\'39.6"W',
        '59º12\'7.7"N 02º15\'39.6"W',
        '59 12\' 7.7" N 02 15\' 39.6" W',
        '59 12\'7.7\'\'N 02 15\'39.6\'\' W',
        '59:12:7.7"N 2:15:39.6W',
        '59 12’7.7’’N 02 15’39.6’’W'
    ];

    var expected = {
        lat: 59 + 12 / 60 + 7.7 / 3600,
        lon: -1 * (2 + 15 / 60 + 39.6 / 3600)
    };

    for (var i = 0; i < testData.length; i ++) {
        t.deepEqual(parseDms(testData[i]), expected, testData[i]);
    }
    
    t.end();
});

test('Correctly parses DMS pairs with hemisphere at beginning', function(t) {

    var testData = [
        'N59°12\'7.7" W02°15\'39.6"',
        'N 59°12\'7.7" W 02°15\'39.6"',
        'N 59.20213888888889° W 2.261°',
        'N 59.20213888888889 W 2.261',
        'W02°15\'39.6" N59°12\'7.7"'
    ];

    var expected = {
        lat: 59 + 12 / 60 + 7.7 / 3600,
        lon: -1 * (2 + 15 / 60 + 39.6 / 3600)
    };

    for (var i = 0; i < testData.length; i ++) {
        t.deepEqual(parseDms(testData[i]), expected, testData[i]);
    }

    t.end();
});

test('Correctly parses different separators between lat / lon pairs', function(t) {

    var testData = [
        '59°12\'7.7"N  02°15\'39.6"W',
        '59°12\'7.7"N , 02°15\'39.6"W',
        '59°12\'7.7"N,02°15\'39.6"W'
    ];

    var expected = {
        lat: 59 + 12 / 60 + 7.7 / 3600,
        lon: -1 * (2 + 15 / 60 + 39.6 / 3600)
    };

    for (var i = 0; i < testData.length; i ++) {
        t.deepEqual(parseDms(testData[i]), expected, testData[i]);
    }

    t.end();
});

test('Will parse a single coordinate with hemisphere', function(t) {

    var testData = [
        '59°12\'7.7"N',
        '02°15\'39.6"W'
    ];

    var expected = [
        {
            lat: 59 + 12 / 60 + 7.7 / 3600,
            lon: undefined
        },{
            lat: undefined,
            lon: -1 * (2 + 15 / 60 + 39.6 / 3600)
        }
    ];


    for (var i = 0; i < testData.length; i ++) {
        t.deepEqual(parseDms(testData[i]), expected[i], testData[i]);
    }

    t.end();
});

test('Will parse a single coordinate with no hemisphere and return a number', function(t) {

    var testData = [
        '59°12\'7.7"',
        '02°15\'39.6"',
        '-02°15\'39.6"'
    ];

    var expected = [
        59 + 12 / 60 + 7.7 / 3600,
        2 + 15 / 60 + 39.6 / 3600,
        -1 * (2 + 15 / 60 + 39.6 / 3600)
    ];


    for (var i = 0; i < testData.length; i ++) {
        t.deepEqual(parseDms(testData[i]), expected[i], testData[i]);
    }

    t.end();

});

test('Will infer first coordinate is lat, second lon, if no hemisphere letter is included', function(t) {

    var testData = [
        '59°12\'7.7" -02°15\'39.6"',
        '59°12\'7.7", -02°15\'39.6"',
    ];

    var expected = {
        lat: 59 + 12 / 60 + 7.7 / 3600,
        lon: -1 * (2 + 15 / 60 + 39.6 / 3600)
    };

    for (var i = 0; i < testData.length; i ++) {
        t.deepEqual(parseDms(testData[i]), expected, testData[i]);
    }

    t.end();

});

test('throws an error for invalid data', function(t) {

    var testData = [
        'Not DMS string'
    ];

    for (var i = 0; i < testData.length; i ++) {
        t.throws(parseDms.bind(null, testData[i]), /Could not parse string/, "Throws for '" + testData[i] + "'");
    }

    t.end();
});

test('Throws for degrees out of range', function(t) {
    t.throws(parseDms.bind(null, '190°12\'7.7" -02°15\'39.6"'), /Degrees out of range/);
    t.end();
});

test('Throws for minutes out of range', function(t) {
    t.throws(parseDms.bind(null, '59°65\'7.7" -02°15\'39.6"'), /Minutes out of range/);
    t.end();
});

test('Throws for seconds out of range', function(t) {
    t.throws(parseDms.bind(null, '59°12\'65.5" -02°15\'39.6"'), /Seconds out of range/);
    t.end();
});

test('Correctly parses DMS with decimal minutes', function(t) {

    var testData = [
        'N59°12.105\' W02°15.66\'',
        'N59:12.105\' W02:15.66\'',
        'N59:12.105 W02:15.66',
        '59:12.105\'N 02:15.66\'W'
    ];

    var expected = {
        lat: 59 + 12.105 / 60,
        lon: -1 * (2 + 15.66 / 60)
    };

    for (var i = 0; i < testData.length; i ++) {
        t.deepEqual(parseDms(testData[i]), expected, testData[i]);
    }

    t.end();
});

test('Correctly parses DMS with no minutes or seconds', function(t) {

    var testData = [
        '59°N 02°W'
    ];

    var expected = {
        lat: 59,
        lon: -2
    };

    for (var i = 0; i < testData.length; i ++) {
        t.deepEqual(parseDms(testData[i]), expected, testData[i]);
    }

    t.end();
});

test('Parse decimal degrees as decimal degrees', function(t) {

    var testData = [
        '51.5, -0.126',
        '51.5,-0.126',
        '51.5 -0.126'
    ];

    var expected = {
        lat: 51.5,
        lon: -0.126
    };

    for (var i = 0; i < testData.length; i ++) {
        t.deepEqual(parseDms(testData[i]), expected, testData[i]);
    }

    t.end();
});

test('Parse DMS with separators and spaces', function(t) {

    var testData = [
        '59° 12\' 7.7" N 02° 15\' 39.6" W',
        '59º 12\' 7.7" N 02º 15\' 39.6" W',
        '59 12’ 7.7’’N 02 15’ 39.6’’W'
    ];

    var expected = {
        lat: 59 + 12 / 60 + 7.7 / 3600,
        lon: -1 * (2 + 15 / 60 + 39.6 / 3600)
    };

    for (var i = 0; i < testData.length; i ++) {
        t.deepEqual(parseDms(testData[i]), expected, testData[i]);
    }
    
    t.end();
});


javascripts/main.js
var $ = require('jquery');
var parseDMS = require('parse-dms');
var csv2geojson = require('csv2geojson');
var tokml = require('tokml');

var mimeTypes = {
    csv: 'text/csv',
    geojson: 'application/json',
    kml: 'application/vnd.google-earth.kml+xml'
};

$('#dms')
    .keyup(onKeyup)
    .focus(onFocus)
    .focus()
    .trigger("keyup");

$('#dec')
    .focus(onFocus);

function onKeyup() {
    var input = this.value.split('\n'),
        output = [],
        errors = [],
        coords;

    $(this).parent().removeClass('has-error');
    $('#error-msg').addClass('hidden');

    for (var i = 0; i < input.length; i++) {
        if (input[i].trim() === "") continue;
        try {
            coords = parseDMS(input[i]);
            output.push(coords.lat + "," + coords.lon);
        } catch (e) {
            errors.push(input[i]);
        }
    }

    $('#dec').val(output.join('\n'));

    if (errors.length) {
        var msg = '<b>Warning:</b> Could not parse input values "' + errors.join('", "') + '"';
        $(this).parent().addClass('has-error');
        $('#error-msg').removeClass('hidden').html(msg);
    }

    if (output.length) {
        $("#download-links").removeClass('hidden');

        var downloads = {};

        downloads.csv = "lat,lon\n" + output.join('\n');

        csv2geojson.csv2geojson(downloads.csv, function(err, data) {
            if (err) downloads['geojson-points'] = "";
            downloads['geojson-points'] = data;
        });

        downloads['geojson-line'] = csv2geojson.toLine(downloads['geojson-points']);
        downloads['geojson-polygon'] = csv2geojson.toPolygon(downloads['geojson-points']);

        downloads['kml-points'] = tokml(downloads['geojson-points']);
        downloads['kml-line'] = tokml(downloads['geojson-line']);
        downloads['kml-polygon'] = tokml(downloads['geojson-polygon']);

        for (var prop in downloads) {
            var type = prop.split('-')[0];

            var content = type === 'geojson' ? JSON.stringify(downloads[prop]) : downloads[prop];

            $('#' + prop + '-download')
                .attr('href','data:' + mimeTypes[type] + ';charset=utf8,' + encodeURIComponent(content));
        }
    } else {
        $('#download-links').addClass('hidden');
    }
}

function onFocus() {
    $(this).select();
}


index.js
'use strict';

module.exports = function(dmsString) {

    dmsString = dmsString.trim();

    // Inspired by https://gist.github.com/JeffJacobson/2955437
    // See https://regex101.com/r/kS2zR1/3
    var dmsRe = /([NSEW])?\s?(-)?(\d+(?:\.\d+)?)[°º:d\s]?\s?(?:(\d+(?:\.\d+)?)['’‘′:]?\s?(?:(\d{1,2}(?:\.\d+)?)(?:"|″|’’|'')?)?)?\s?([NSEW])?/i;

    var result = {};

    var m1, m2, decDeg1, decDeg2, dmsString2;

    m1 = dmsString.match(dmsRe);

    if (!m1) throw 'Could not parse string';

    // If dmsString starts with a hemisphere letter, then the regex can also capture the 
    // hemisphere letter for the second coordinate pair if also in the string
    if (m1[1]) {
        m1[6] = undefined;
        dmsString2 = dmsString.substr(m1[0].length - 1).trim();
    } else {
        dmsString2 = dmsString.substr(m1[0].length).trim();
    }

    decDeg1 = decDegFromMatch(m1);

    m2 = dmsString2.match(dmsRe);

    decDeg2 = m2 ? decDegFromMatch(m2) : {};

    if (typeof decDeg1.latLon === 'undefined') {
        if (!isNaN(decDeg1.decDeg) && isNaN(decDeg2.decDeg)) {
            // If we only have one coordinate but we have no hemisphere value,
            // just return the decDeg number
            return decDeg1.decDeg;
        } else if (!isNaN(decDeg1.decDeg) && !isNaN(decDeg2.decDeg)) {
            // If no hemisphere letter but we have two coordinates,
            // infer that the first is lat, the second lon
            decDeg1.latLon = 'lat';
            decDeg2.latLon = 'lon';
        } else {
            throw 'Could not parse string';
        }
    }

    // If we parsed the first coordinate as lat or lon, then assume the second is the other
    if (typeof decDeg2.latLon === 'undefined') {
        decDeg2.latLon = decDeg1.latLon === 'lat' ? 'lon' : 'lat';
    }

    result[decDeg1.latLon] = decDeg1.decDeg;
    result[decDeg2.latLon] = decDeg2.decDeg;

    return result;

};

function decDegFromMatch(m) {
    var signIndex = {
        "-": -1,
        "N": 1,
        "S": -1,
        "E": 1,
        "W": -1
    };

    var latLonIndex = {
        "N": "lat",
        "S": "lat",
        "E": "lon",
        "W": "lon"
    };
    var degrees, minutes, seconds, sign, latLon;

    sign = signIndex[m[2]] || signIndex[m[1]] || signIndex[m[6]] || 1;
    degrees = Number(m[3]);
    minutes = m[4] ? Number(m[4]) : 0;
    seconds = m[5] ? Number(m[5]) : 0;
    latLon = latLonIndex[m[1]] || latLonIndex[m[6]];

    if (!inRange(degrees, 0, 180)) throw 'Degrees out of range';
    if (!inRange(minutes, 0, 60)) throw 'Minutes out of range';
    if (!inRange(seconds, 0, 60)) throw 'Seconds out of range';

    return {
        decDeg: sign * (degrees + minutes / 60 + seconds / 3600),
        latLon: latLon
    };
}

function inRange(value, a, b) {
    return value >= a && value <= b;
}
